In [134]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [135]:
df = pd.read_csv('Week 9 Data Set/bank-additional/bank-additional-full.csv', sep=";")

In [136]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

In [137]:
df.dtypes

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object

In [138]:
df.isna().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

## While there are no NA values existing in the dataset, there are 'unknown' values listed in the data documentation. For each of these columns, we can make another column that identifies whether the row has a unknown value or not with a '0' or '1'

In [139]:
# Create new columns for unknown values
df['job_unknown'] = 0
df['marital_unknown'] = 0
df['education_unknown'] = 0
df['default_unknown'] = 0
df['housing_unknown'] = 0
df['loan_unknown'] = 0

# Set new columns to 1 for rows where value is unknown
df.loc[df['job'] == 'unknown', 'job_unknown'] = 1
df.loc[df['marital'] == 'unknown', 'marital_unknown'] = 1
df.loc[df['education'] == 'unknown', 'education_unknown'] = 1
df.loc[df['default'] == 'unknown', 'default_unknown'] = 1
df.loc[df['housing'] == 'unknown', 'housing_unknown'] = 1
df.loc[df['loan'] == 'unknown', 'loan_unknown'] = 1

## Now we can check our output variable to see if there is an imbalance in the data

In [140]:
# Now checking our target variable
df['y'].value_counts()

no     36548
yes     4640
Name: y, dtype: int64

# There is an imbalance in our target variables, with 36548 "no" data points and 4640 "yes" data points. 

### Specifically, the "no" and "yes" is if the client subscribed a term deposit. This is what we want our model to predict.

### We will balance the data for modeling, with two seperate categories: datasets balanced with the duration variable, and datasets balanced without. This is because the duration variable is difficult for a business to draw insights from in terms of campaigning.

In [141]:
df['duration'].value_counts()

90      170
85      170
136     168
73      167
124     164
       ... 
1569      1
1053      1
1263      1
1169      1
1868      1
Name: duration, Length: 1544, dtype: int64

In [142]:
filtered_duration_df = df[df['duration'] < 10]
filtered_duration_df['y'].value_counts()

no    290
Name: y, dtype: int64

In [143]:
filtered_duration_df = df[df['duration'] < 50]
filtered_duration_df['y'].value_counts()

no     3101
yes       1
Name: y, dtype: int64

In [144]:
filtered_duration_df = df[df['duration'] < 75]
filtered_duration_df['y'].value_counts()

no     6164
yes      12
Name: y, dtype: int64

In [145]:
filtered_duration_df = df[df['duration'] < 100]
filtered_duration_df['y'].value_counts()

no     9804
yes      78
Name: y, dtype: int64

In [146]:
filtered_duration_df = df[df['duration'] < 125]
filtered_duration_df['y'].value_counts()

no     13349
yes      179
Name: y, dtype: int64

In [147]:
filtered_duration_df = df[df['duration'] < 150]
filtered_duration_df['y'].value_counts()

no     16573
yes      327
Name: y, dtype: int64

In [148]:
filtered_duration_df = df[df['duration'] < 200]
filtered_duration_df['y'].value_counts()

no     21946
yes      706
Name: y, dtype: int64

In [149]:
filtered_duration_df = df[df['duration'] < 250]
filtered_duration_df['y'].value_counts()

no     25751
yes     1116
Name: y, dtype: int64

In [150]:
filtered_duration_df = df[df['duration'] < 300]
filtered_duration_df['y'].value_counts()

no     28429
yes     1509
Name: y, dtype: int64

In [151]:
filtered_duration_df = df[df['duration'] < 400]
filtered_duration_df['y'].value_counts()

no     31898
yes     2091
Name: y, dtype: int64

In [152]:
filtered_duration_df = df[df['duration'] < 38]
filtered_duration_df['y'].value_counts()

no     2161
yes       1
Name: y, dtype: int64

In [153]:
filtered_duration_df = df[df['duration'] < 37]
filtered_duration_df['y'].value_counts()

no    2092
Name: y, dtype: int64

In [154]:
filtered_duration_df = df[df['duration'] > 36]
filtered_duration_df['y'].value_counts()

no     34456
yes     4640
Name: y, dtype: int64

## Summarization: all data points below 36 seconds in duration column result in "no". The amount of "yes" answers increase dramatically past 100 seconds. 

It may be possible to remove all data points below 36 seconds (or other threshold) when modeling to reduce noise.

## We can perform more transformations after we perform the Exploratory Data Analysis. This is a way to prepare our data in the future for modeling.

In [155]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  cons.conf.idx  euribor3m  nr.employed   y  \
0   may         mon  ...          -36.4      4.857       5191.0  no   
1   may         mon  ...          -36.4      4.857       5191.0  no   
2   may         mon  ...          -36.4      4.857       5191.0  no   
3   may         mon  ...          -36.4      4.857       5191.0  no   
4   may         mon  ...          -36.4      4.857       5191.0  no   

  job_unknown  marital_unknown  education_unknown  default_unknown  \
0           0                0                  0                0   
1           0                0                  0                1   
2           0                0                  0                0   
3           0                0                  0                0   
4           0                0                  0                0   

   housing_unknown  loan_unknown  
0                0             0  
1                0             0  
2                0             0  
3                0             0  
4                0             0  

[5 rows x 27 columns]

In [156]:
# Make a filtered csv file with all data points where duration is above 36 seconds
filtered_duration_df.to_csv('filtered_duration_bank', index=False)

In [159]:
# Make a new csv file with our new columns. 
binary_col_bank_full = df.to_csv('binary_col_bank_full', index=False)